apps/prairielearn/src/ee/lib/validateHTML.ts

```diff
@@ -0,0 +1,616 @@
+  'tabindex',
+  'title',
+  'translate',
+];
+
+/**
+ * Checks that the required property is an int (or property to be inserted) or adds an error to the provided list.
+ * @param tag The name of the tag being checked.
+ * @param key The property name.
+ * @param val The property value.
+ * @param errors The list of errors to add to.
+ */
+function assertInt(tag: string, key: string, val: string, errors: string[]) {
+  if (!(/^\d+$/.test(val) || /^\{\{.*\}\}$/.test(val))) {
+    errors.push(
+      `${tag}: value for attribute ${key} must be an integer, but value provided is "${val}"`,
+    );
+  }
+}
+
+/**
+ * Checks that the required property is an float (or property to be inserted) or adds an error to the provided list.
+ * @param tag The name of the tag being checked.
+ * @param key The property name.
+ * @param val The property value.
+ * @param errors The list of errors to add to.
+ */
+function assertFloat(tag: string, key: string, val: string, errors: string[]) {
+  if (!(/^(\d+)\.?(\d*)(e-\d+)?$/.test(val) || /^\{\{.*\}\}$/.test(val))) {
+    errors.push(
+      `${tag}: value for attribute ${key} must be an floating-point number, but value provided is "${val}"`,
+    );
+  }
+}
+
+/**
+ * Checks that the required property is in a list of possibilities (or property to be inserted) or adds an error to the provided list.
+ * @param tag The name of the tag being checked.
+ * @param key The property name.
+ * @param val The property value.
+ * @param choices The list of potential choices for the property.
+ * @param errors The list of errors to add to.
+ */
+function assertInChoices(
+  tag: string,
+  key: string,
+  val: string,
+  choices: string[],
+  errors: string[],
+) {
+  if (!(choices.includes(val) || /^\{\{.*\}\}$/.test(val))) {
+    errors.push(
+      `${tag}: value for attribute ${key} must be in ${choices}, but value provided is ${val}`,
+    );
+  }
+}
+
+/**
+ * Checks that the required property is a boolean (or property to be inserted) or adds an error to the provided list.
+ * @param tag The name of the tag being checked.
+ * @param key The property name.
+ * @param val The property value.
+ * @param choices The list of potential choices for the property.
+ * @param errors The list of errors to add to.
+ */
+function assertBool(tag: string, key: string, val: string, errors: string[]) {
+  assertInChoices(tag, key, val, ['true', 'false'], errors);
+}
+
+/**
+ * Checks that a tag has valid properties.
+ * @param ast The tree to consider, rooted at the tag.
+ * @returns The list of errors for the tag, if any.
+ */
+function checkTag(ast: any): string[] {
+  switch (ast.tagName) {
+    case 'pl-multiple-choice':
+      return checkMC(ast);
+    case 'pl-integer-input':
+      return checkInt(ast);
+    case 'pl-number-input':
+      return checkNum(ast);
+    case 'pl-answer':
+      return checkMCAnswer(ast);
+    case 'pl-overlay':
+      return checkOverlay(ast);
+    case 'pl-location':
+      return checkLocation(ast);
```

Comment:
Are these really frequently used enough to make the cut for manual validation, and do we even expect LLMs to be able to do anything useful with them? I'm surprised to see this here and not e.g. `pl-string-input`.

Reply:
I included all elements that were used in the example course that could be created using the current generation system upon a cursory glance at them. I'll include pl-string-input as well (which should probably be included anyways even if there's no example questions).

Reply:
Added `pl-string-input` validation in c0617cd.